flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/mailbox/TaskMailboxImpl.java

```diff
@@ -251,4 +280,79 @@ public void quiesce() {
 	public State getState() {
 		return state;
 	}
+
+	@Override
+	public Mailbox getDownstreamMailbox(AbstractStreamOperator<?> operator) {
+		return getDownstreamMailbox(operator.getOperatorConfig().getChainIndex());
```

Comment:
Suggestion: implementation wise, only the `AbstractStreamOperator<?>.getOperatorConfig().getChainIndex()` is used here.

You can decouple `TaskMailboxImpl` from `AbstractStreamOperator` at all, so the method becomes `Mailbox getDownstreamMailbox(int priority)`.

Reply:
That's what I had in the first place.
However, in respect to a possible branched operator chain (Piotr mentioned that this may come), I wanted to make the priority and its type an implementation detail.
Nevertheless, I see your point about the cyclic coupling and would revert back to int priority. 
Alternatively, we could also take the StreamConfig to avoid the cyclic dependency graph.

Reply:
+1 for the @1u0 comment.

Dependency on `AbstractStreamOperator` is a very heavy one and will make testing/reusing difficult. I would like to avoid it even with potential future changes if `int priority` proves to be not enough. Keep in mind that we can always refactor this in the future and using now `int priority` will not cause us much (if any) overhead in the future if we need to change it.

Reply:
Using int priority everywhere now.